### PR TITLE
Add P1 => P2 MongoDB Upgrade Guide

### LINKED ISSUE #2767: MDBGA: Document an upgrade path for P1 MongoDB users (Epic 3)
- Does introspection work for P1 users?
- Did P1 make any assumptions about relations that are invalid for P2 users?

### PR DESCRIPTION
**Pull request by matthewmueller**
## Describe this PR

This PR provides a short upgrade guide for those trying to migrate from Prisma 1 running on MongoDB to Prisma 2.

Closes https://github.com/prisma/docs/issues/2767

Once this is merged we can merge: https://github.com/prisma/pris.ly/pull/47

### CODE REVIEW

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+The database that stores application data in Prisma 1 is called `default_default`, so we'll add that to the end of the connection string and update the `DATABASE_URL` key in the `.env` file
+
+```bash
+DATABASE_URL="mongodb://prisma:prisma@localhost:27017/default_default"
+```
+
+## Introspect your MongoDB Database
+
+You're now ready to pull the structure of your database down into your Prisma Schema
```
Comment (tanberry) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]:
```suggestion
You're now ready to pull the structure of your database down into your Prisma Schema.
```

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+```bash
+$ npx prisma db pull
+```
+
+And you should see your Prisma Schema in `prisma/schema.prisma` populated with your models.
```
Comment (tanberry) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]:
```suggestion
And you should see your Prisma schema in `prisma/schema.prisma` populated with your models.
```

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+<Admonition type="warning">
+
+If you see the following error: `Error in connector: SCRAM failure: Authentication failed.`, try adding `?authSource=admin` to the end of your connection string and trying again.
+
+</Admonition>
+
+If you run into any other errors with introspection, please refer to this documentation.
```
Comment (tanberry) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]: By "this document" do you mean further down the page, or is it another doc and you forgot to add a link? :-)

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+Take caution in doing these renames because you need to make sure the Prisma Schema still maps properly to the underlying database collections and field names.
+
+Unlike SQL databases, MongoDB doesn't have an explicit understanding of relationships between data. This means that Prisma's introspection is unable to infer those relationships for you.
+
+However, where Prisma 1 stores foreign keys is different than where Prisma 2 expects foreign keys, so if you want to take advantage of relationships, you'll need to shift where the foreign keys are on your database before adding the relationships.
+
+<Admonition type="tip">
+
+💡 Download the Prisma 2 VSCode Extension to provide autocomplete and helpful error messages as you transition your Prisma Schema
```
Comment (tanberry) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]:
```suggestion
💡 Download the Prisma 2 VSCode Extension to provide autocomplete and helpful error messages as you transition your Prisma schema.
```

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+</Admonition>
+
+## Generating a Prisma Client
+
+With the Prisma Schema populated with the schema of your data, you're now ready to generate a Typescript Client to read and write to your MongoDB database
```
Comment (tanberry) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]:
```suggestion
With the Prisma schema populated with the schema of your data, you're now ready to generate a Typescript Client to read and write to your MongoDB database.
```

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+  })
+  console.log(user)
+}
+
+main()
+  .catch(console.error)
+  .finally(() => prisma.$disconnect())
+```
+
+And you should see a user has been created.
```
Comment (tanberry) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]:
```suggestion
And you should see a user was created.
```

**Review by tanberry — COMMENTED** [review]
Looks great, and thanks a ton for writing this whole guide.

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx [hunk 7]
```diff
@@ -0,0 +1,231 @@
+---
+title: 'Upgrade from MongoDB Beta'
+metaTitle: 'Upgrade from the Prisma 1 MongoDB Beta to Prisma 2'
+metaDescription: 'Learn how to upgrade your MongoDB application running Prisma 1 to Prisma 2.'
+---
+
+## Introduction
+
+This guide aims to help you migrate from the MongoDB Beta on Prisma 1 to MongoDB on Prisma 2. To learn more about the differences between Prisma 1 and Prisma 2, you can refer to [this document](https://www.prisma.io/docs/guides/upgrade-guides/upgrade-from-prisma-1/how-to-upgrade#main-differences-between-prisma-1-and-prisma-version-2x-and-later).
```
Comment (janpio) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]: This is messy, but exact:
```suggestion
This guide aims to help you migrate from the Prisma 1 MongoDB Beta to MongoDB on Prisma 3.12.0 and later. To learn more about the differences between Prisma 1 and Prisma 2.x and later, you can refer to [this document](https://www.prisma.io/docs/guides/upgrade-guides/upgrade-from-prisma-1/how-to-upgrade#main-differences-between-prisma-1-and-prisma-version-2x-and-later).
```

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+
+The scope of this guide is to give you the workflow necessary to perform the migration and highlight some of the gotchas you may encounter.
+
+We unfortunately can't cover all possible scenarios or changes required, but this guide should help you on your journey. Join [our community Slack](https://slack.prisma.io/) or create an issue [on Github](https://github.com/prisma/prisma1/issues/new/choose) with any questions.
+
+Please perform this migration on your staging environment before trying this in production!
```
Comment (janpio) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]: Warning admonition?

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+
+## Requirements
+
+- Must be running MongoDB 4.2+ as a replica set (MongoDB Atlas does this for you automatically)
+- Node.js 12.6.x or 14.x or 16.x
+- TypeScript 4.1.x
+
+## Installing Prisma 2
```
Comment (janpio) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]:
```suggestion
## Installing Prisma 3.12.0 or later
```

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+
+In your project directory run the following commands:
+
+```bash
+$ npm install prisma
```
Comment (janpio) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]: Needed as they could still have a `prisma` in their dependencies already, even pinned:
```suggestion
$ npm install prisma@latest
```

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+$ npx prisma init --datasource-provider=mongodb
+```
+
+This should create the following files:
+
+- `prisma/schema.prisma`: An initial Prisma 2 Schema
```
Comment (janpio) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]:
```suggestion
- `prisma/schema.prisma`: An initial Prisma schema
```

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+Please try again in a project that is not yet using Prisma.
+```
+
+You have likely a `prisma/` directory in your project already. Rename that directory to something like `_prisma/` and try again
+
+</Admonition>
+
+## Find the Connection String to your MongoDB Database
+
+Next you'll want to find the connection string to your MongoDB database. You should be able to find it in your docker-compose file or on Atlas. It's what you'd pass to MongoDB Compass. The connection string should look something like this:
```
Comment (janpio) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]:
```suggestion
Next you'll want to find the connection string to your MongoDB database. You should be able to find it in your `docker-compose.yml` file or on MongoDB Atlas. It's what you'd pass to MongoDB Compass. The connection string should look something like this:
```

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+
+```bash
+mongodb://<user>:<pass>@<host>:27017
+```
+
+The database that stores application data in Prisma 1 is called `default_default`, so we'll add that to the end of the connection string and update the `DATABASE_URL` key in the `.env` file
+
+```bash
```
Comment (janpio) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]:
````suggestion
```bash file=.env
````

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+
+If you see the following error: `Error in connector: SCRAM failure: Authentication failed.`, try adding `?authSource=admin` to the end of your connection string and trying again.
+
+</Admonition>
+
+If you run into any other errors with introspection, please refer to this documentation.
+
+## Touching up your Prisma Schema
+
+The generated Prisma Client from a freshly introspected MongoDB database doesn't have the best API. You can adjust the names and use `@map` and `@@map` to map to the underlying collection and field names:
```
Comment (janpio) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]:
```suggestion
The generated Prisma Client from a freshly introspected Prisma 1 based MongoDB database doesn't have the best API. You can adjust the names and use `@map` and `@@map` to map to the underlying collection and field names:
```

Comment (janpio) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]: Is that always the case?

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+💡 Download the Prisma 2 VSCode Extension to provide autocomplete and helpful error messages as you transition your Prisma Schema
+
+</Admonition>
+
+## Generating a Prisma Client
+
+With the Prisma Schema populated with the schema of your data, you're now ready to generate a Typescript Client to read and write to your MongoDB database
+
+```bash
+$ prisma generate
```
Comment (janpio) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]:
```suggestion
$ npx prisma generate
```

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+  const posts = await prisma.post.findMany()
+  console.log(posts)
+}
+
+main()
+  .catch(console.error)
+  .finally(() => prisma.$disconnect())
+```
+
+Make sure `ts-node` is installed and run:
```
Comment (janpio) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]:
```suggestion
Make sure `ts-node` is installed globally and run:
```

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+```
+
+And you should see a user has been created.
+
+<Admonition type="warning">
+
+If you see the following error:
+
+```
+Prisma needs to perform transactions, which requires your MongoDB server to be run as a replica set. [https://pris.ly/d/mongodb-replica-set](https://pris.ly/d/mongodb-replica-set)
```
Comment (janpio) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]: No links in code blocks:
```suggestion
Prisma needs to perform transactions, which requires your MongoDB server to be run as a replica set. https://pris.ly/d/mongodb-replica-set
```

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+```
+
+This means that your MongoDB database isn't running as a replica set. Refer to the link above for steps to resolve this issue.
```
Comment (janpio) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]:
```suggestion
This means that your MongoDB database isn't running as a replica set. Refer to [the link above](https://pris.ly/d/mongodb-replica-set) for steps to resolve this issue.
```

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx (hunk 3 above)
Comment (matthewmueller) [reply content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]: Good catch! This should link to @keerlu's guide.

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx (hunk 8 above)
Comment (matthewmueller) [reply content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]: Good idea

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx (hunk 14 above)
Comment (matthewmueller) [reply content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]: You mean: _doesn't_ should probably be _may not_? Just re-wrote this paragraph.

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx (hunk 3 above)
Comment (matthewmueller) [reply content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/08-upgrade-from-mongodb-beta.mdx]: Actually, I realized there are no docs on how to handle introspection errors, so I removed this.

**Comment by matthewmueller** [comment]
Just waiting on #3017.

Merge this one right after: https://github.com/prisma/pris.ly/pull/47

**Pull request merged**